File: book/python_algorithm_interview/js/04.ts
function solution(paragraph: string, banned: string[]) {
  //   let cleaned = paragraph
  //     .split(" ")
  //     .map((item) => item.replace(/[^\w]/, ""))
  //     .join(" ")
  //     .toLowerCase();

  let cleaned = "";

  for (let w of paragraph) {
    cleaned += w.replace(/[^\w]/, " ").toLowerCase();
  }

  let count = {};
  let max = { key: "", value: 0 };

  for (let w of cleaned.split(" ").filter((item) => !banned.includes(item))) {
    if (!w || w in banned) continue;

    if (w in count) {
      count[w] += 1;
    } else {
      count[w] = 1;
    }
  }
  console.log("count: ", count);
  for (let c in count) {
    if (count[c] > max.value) {
      console.log("here: ", c, count[c]);
      max = { key: c, value: count[c] };
    }
  }

  return max.key;
}

console.log(solution("Bob. hIt, baLl", ["bob", "hit"]));

File: achieve/book/python_algorithm_interview/Q/04.py
from typing import List
import collections
import re


def format_isalnum(s: str) -> str:
    result = ''
    for item in s:
        if item.isalnum():
            result += item
        else:
            result += ' '
    return result


def most_common_word(s: str, banned: List[str]) -> str:
    counter = {}
    lowered = s.lower()
    formatted = format_isalnum(lowered)

    for item in formatted.split():
        if item != banned[0]:
            if item in counter:
                counter[item] += 1
            else:
                counter[item] = 1

    # 아래 방법으로는 제대로 카운트 되지 않음
    # return max(counter, key=counter.get) 와 같은 방법으로 카운트 해야 함.
    max_idx = max(counter.values())
    result = ''

    for i, v in enumerate(counter):
        if i == max_idx:
            result = v

    return result


def most_common_word_refactor(paragraph: str, banned: List[str]) -> str:
    counter = collections.defaultdict(int)
    lowered = paragraph.lower()
    formatted = format_isalnum(lowered)
    for item in formatted.split():
        if item not in banned:
            counter[item] += 1
    return max(counter, key=counter.get)


def most_common_word_book(paragraph: str, banned: List[str]) -> str:
    words = [word for word in re.sub(
        r'[^\w]', ' ', paragraph).lower().split() if word not in banned]
    counts = collections.Counter(words)
    return counts.most_common(1)[0][0]


paragraph = "Bob hit a ball, the hit BALL flew far after it was hit."
banned = ["hit"]
print(most_common_word_book(paragraph, banned))
